core/invoker/src/main/scala/whisk/core/dispatcher/ActivationFeed.scala

```diff
@@ -103,14 +110,29 @@ protected class ActivationFeed(
             } else logging.debug(this, "dropping fill request until feed is drained")
 
         case n: ActivationNotification =>
+            downstreamCapacity += 1
             pipelineOccupancy -= 1
             logging.info(this, s"received ActivationNotification: $n / pipelineOccupancy=$pipelineOccupancy / pipelineFillThreshold=$pipelineFillThreshold")
             if (pipelineOccupancy < 0) {
                 logging.error(this, "pipelineOccupancy<0")
             }
+            send()
             fill()
     }
 
+    /** Send as many messages as we're allowed to, or as we have depending on what happens first. */
+    @tailrec
+    private def send(): Unit = {
+        if (downstreamCapacity > 0 && outstanding.size > 0) {
+            downstreamCapacity -= 1
```

Comment:
i think this is missing pipeline occupancy update (fill, then send all messages at once).